### PR TITLE
quantlib: fix configure options

### PR DESCRIPTION
**Pull request by yellowbunchu**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Does your submission pass `brew audit --strict --online <formula>` (after doing `brew install <formula>`)?
### Description

I was having linking problems trying to use quantlib-swig, figured out it is due to wrong configure options and environment flags. 
Changing these as described on [quantlib page](http://quantlib.org/install/macosx.shtml) fixed the issue.

### CODE REVIEW

File: Formula/quantlib.rb [hunk 1]
```diff
@@ -38,12 +37,20 @@ class Quantlib < Formula
   def install
     ENV.cxx11 if build.cxx11?
     if build.head?
-      Dir.chdir "QuantLib"
       system "./autogen.sh"
     end
+    if MacOS.version >= :mavericks
+      ENV.append "LDFLAGS", "-stdlib=libstdc++ -mmacosx-version-min=10.6"
+      ENV.append "CXXFLAGS", "-stdlib=libstdc++ -mmacosx-version-min=10.6"
```
Comment (ilovezfs) [inline Formula/quantlib.rb]: Why are these needed?

File: Formula/quantlib.rb [hunk 2]
```diff
@@ -38,12 +37,20 @@ class Quantlib < Formula
...
+    end
     system "./configure", "--disable-dependency-tracking",
+                          "--enable-static",
+                          "--with-boost-include=/usr/local/include/",
+                          "--with-boost-lib=/usr/local/lib/",
```
Comment (ilovezfs) [inline Formula/quantlib.rb]: These paths cannot be hard coded.  You can use `Formula["boost"].opt_lib` and `.opt_include`

File: Formula/quantlib.rb (hunk 1 above)
Comment (aliyakamercan) [inline Formula/quantlib.rb]: I don't know
quantlib website has:
> Users have reported linking problems since Mac OS X 10.9; the solution (thanks to Albert Azout for pointing it out) seems to be to set the environment flags CXXFLAGS and LDFLAGS to -stdlib=libstdc++ -mmacosx-version-min=10.6 before compiling.

**Comment by ilovezfs** [comment]
Were you actually having a problem building?

This is without any changes:

```
iMac-TMP:Formula joe$ brew install --build-from-source quantlib
==> Downloading https://downloads.sourceforge.net/project/quantlib/QuantLib/1.7/QuantLib-1.7.tar.gz
==> Downloading from http://iweb.dl.sourceforge.net/project/quantlib/QuantLib/1.7/QuantLib-1.7.tar.gz
######################################################################## 100.0%
==> Downloading https://gist.githubusercontent.com/enricodetoma/7d7b137b69726815f070/raw/aa952c28854d
######################################################################## 100.0%
==> Patching
==> Applying patch-FRA
patching file Examples/FRA/Makefile.am
patching file Examples/FRA/Makefile.in
==> ./configure --prefix=/usr/local/Cellar/quantlib/1.7 --with-lispdir=/usr/local/Cellar/quantlib/1.7
==> make install
==> Caveats
Emacs Lisp files have been installed to:
  /usr/local/share/emacs/site-lisp/quantlib
==> Summary
🍺  /usr/local/Cellar/quantlib/1.7: 1,264 files, 149.7M, built in 11 minutes 39 seconds
iMac-TMP:Formula joe$ brew test quantlib
Testing quantlib
==> Using the sandbox
==> /usr/local/Cellar/quantlib/1.7/bin/quantlib-config --prefix=/usr/local/Cellar/quantlib/1.7 --libs

```

Incidentally, the current version is 1.7.1 not 1.7, so maybe it was an issue with the newer version?

**Comment by ilovezfs** [comment]
Also, some good news with 1.7.1: the patch in the formula is in the new release, so it can be removed alongside the version bump.

File: Formula/quantlib.rb (hunk 1 above)
Comment (UniqMartin) [inline Formula/quantlib.rb]: No matter what the linking problems are (or were), this is definitely not a solution but more of a dirty hack that is going to create new issues sooner than later. If the original issue still persists, it would be good to understand the root cause and fix it properly instead of perpetuating a more than two year old workaround …

**Comment by aliyakamercan** [comment]
@ilovezfs it was building fine and tests were passing. The problem occurred when I tried to use https://github.com/lballabio/QuantLib-SWIG and build for python. I would get errors 
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/valarray:4035:59: error: 
      'value_type' is a private member of 
      'boost::iterators::detail::operator_brackets_proxy<QuantLib::step_iterator<double 
      *> >' 
    __val_expr<_BinaryOp<__bit_shift_left<typename _Expr::value_type>, 
                                                          ^ 
quantlib_wrap.cpp:8230:23: note: while substituting deduced template arguments 
      into function template 'operator<<' [with _Expr = 
      boost::iterators::detail::operator_brackets_proxy<QuantLib::step_iterator<double 
      *> >] 
                    s << (*self)[i][j]; 
                      ^ 
In file included from quantlib_wrap.cpp:2649: 
In file included from /usr/local/Cellar/quantlib/1.6.1/include/ql/quantlib.hpp:47: 
In file included from /usr/local/Cellar/quantlib/1.6.1/include/ql/math/all.hpp:35: 
In file included from /usr/local/Cellar/quantlib/1.6.1/include/ql/math/matrixutilities/all.hpp:4: 
In file included from /usr/local/Cellar/quantlib/1.6.1/include/ql/math/matrixutilities/basisincompleteordered.hpp:25: 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/valarray:4036:46: error: 
      'value_type' is a private member of 
      'boost::iterators::detail::operator_brackets_proxy<QuantLib::step_iterator<double 
      *> >' 
               __scalar_expr<typename _Expr::value_type>, _Expr> > 
```

File: Formula/quantlib.rb (hunk 1 above)
Comment (aliyakamercan) [inline Formula/quantlib.rb]: I think `stdlib=libstdc++` is because libc++ became default by mavericks. I am not so sure why ` -mmacosx-version-min=10.6` is needed. I didn't test if it would work w/o it.

Quantlib advertises this solution. Currently, this formula is virtually useless.

Comment (mikemcquaid) [inline Formula/quantlib.rb]: We need to use libc++ unfortunately. Might be worth filing an issue upstream and removing this project if we can't get it working.

**Comment by ilovezfs** [comment]
@aliyakamercan That's actually fixed if you build from the HEAD of both repositories.

**Comment by ilovezfs** [comment]
https://github.com/lballabio/QuantLib-SWIG/commit/b14a50c380fd5c3171ff0d005253ba7831aa2e69
https://github.com/lballabio/QuantLib-SWIG/pull/4

**Comment by aliyakamercan** [comment]
@ilovezfs that's great, should we close this then ?

**Comment by ilovezfs** [comment]
@aliyakamercan I think so. Would you like to open a new one where the only change is the version bump?

**Comment by ilovezfs** [comment]
@aliyakamercan btw, I hacked together a quantlib-swig formula
```
class QuantlibSwig < Formula
  desc "QuantLib extension modules"
  homepage "https://github.com/lballabio/QuantLib-SWIG"
  url "https://github.com/lballabio/QuantLib-SWIG/archive/QuantLib-SWIG-v1.7.tar.gz"
  sha256 "88c3738ad136e5988bc9a9b229b6f3f3f3d486c66ca23052b661f46ac3bdcc1d"
  head "https://github.com/lballabio/QuantLib-SWIG.git"

  depends_on "autoconf" => :build
  depends_on "automake" => :build
  depends_on "libtool" => :build
  depends_on "swig" => :build
  depends_on "quantlib"

  patch do
    url "https://github.com/lballabio/QuantLib-SWIG/pull/4.patch"
    sha256 "e7ed9adad78c0b5806b6f04e165593292b2d57c30a7195835c5e915c630b6ed1"
  end

  def install
    system "./autogen.sh"
    system "./configure",
           "--prefix=#{prefix}",
           "--disable-guile",
           "--disable-java"

    system "make"
    system "make", "install"
  end
end
```

Remaining issues with it are getting the guile and java components to build, and adjust the the ruby and python directories to be sandbox compliant. It currently tries to write to
```
/Library/Python/2.7/site-packages
/Library/Ruby/Site/2.0.0/universal-darwin15/
/Library/Ruby/Site/2.0.0/
```

**Comment by ilovezfs** [comment]
Closing since the issue the PR seeks to address has been fixed by changes upstream to the related `quantlib-swig`.
